File: src/main/java/com/tmvkrpxl0/tcombat/common/network/packets/TCombatPacketHandler.java
package com.tmvkrpxl0.tcombat.common.network.packets;

import com.tmvkrpxl0.tcombat.TCombatMain;
import com.tmvkrpxl0.tcombat.common.skills.AbstractActiveSkill;
import com.tmvkrpxl0.tcombat.common.skills.AbstractSkill;
import com.tmvkrpxl0.tcombat.common.util.TCombatUtil;
import net.minecraft.entity.Entity;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.player.ServerPlayerEntity;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.Util;
import net.minecraft.util.text.StringTextComponent;
import net.minecraft.world.World;
import net.minecraftforge.fml.network.NetworkRegistry;
import net.minecraftforge.fml.network.simple.SimpleChannel;
import net.minecraftforge.fml.server.ServerLifecycleHooks;

import java.util.LinkedList;
import java.util.UUID;

public class TCombatPacketHandler {
    private static final String PROTOCOL_VERSION = "1";
    public static final String CHANNEL_NAME = "tcombat_packet_handler";
    public static final SimpleChannel INSTANCE = NetworkRegistry.newSimpleChannel(
            new ResourceLocation(TCombatMain.MODID, CHANNEL_NAME),
            () -> PROTOCOL_VERSION,
            PROTOCOL_VERSION::equals,
            PROTOCOL_VERSION::equals
    );
    public TCombatPacketHandler(){
        int id = 0;
        INSTANCE.registerMessage(id, SkillRequestPacket.class,
                (skillRequestPacket, packetBuffer) ->
                        packetBuffer.writeRegistryId(skillRequestPacket.getSkill()),
                packetBuffer -> {
                    AbstractSkill skill = packetBuffer.readRegistryIdSafe(AbstractSkill.class);
                    if(!(skill instanceof AbstractActiveSkill)){
                        throw new IllegalArgumentException("Only Active Skills can be sent!!!");
                    }
                    return new SkillRequestPacket((AbstractActiveSkill) skill);
                }, (skillRequestPacket, contextSupplier) -> {
                    contextSupplier.get().enqueueWork(() -> skillRequestPacket.getSkill().execute(contextSupplier.get().getSender()));
                    contextSupplier.get().setPacketHandled(true);
                });
        id++;
        INSTANCE.registerMessage(id, TargetSetPacket.class, (targetSetPacket, packetBuffer) -> {
            packetBuffer.writeUUID(targetSetPacket.getUniqueID());
            packetBuffer.writeVarIntArray(targetSetPacket.entityIds());
        },
                packetBuffer -> {
                    UUID uuid = packetBuffer.readUUID();
                    int[] entityIds = packetBuffer.readVarIntArray(100);
                    ServerLifecycleHooks.getCurrentServer().getPlayerList().getPlayer(uuid)
                            .sendMessage(new StringTextComponent("Sending.. Count: " + entityIds.length), Util.NIL_UUID);
                    return new TargetSetPacket(uuid, entityIds);
                }, (targetSetPacket, contextSupplier) -> {
                    LinkedList<LivingEntity> list = new LinkedList<>();
                    ServerPlayerEntity playerEntity = contextSupplier.get().getSender();
                    World world = playerEntity.level;
                    contextSupplier.get().enqueueWork(() -> {
                        for(int i : targetSetPacket.entityIds()){
                            Entity entity = world.getEntity(i);
                            if(entity instanceof LivingEntity)list.add((LivingEntity)entity);
                        }
                        TCombatUtil.setTargets(playerEntity, list);
                        playerEntity.sendMessage(new StringTextComponent("Succeed! Count: " + TCombatUtil.getTargets(playerEntity).size()), Util.NIL_UUID);
                    });
                    contextSupplier.get().setPacketHandled(true);
                });
    }
}

File: src/main/kotlin/com/tmvkrpxl0/tcombat/common/network/packets/TCombatPacketHandler.kt
package com.tmvkrpxl0.tcombat.common.network.packets

import com.tmvkrpxl0.tcombat.TCombatMain
import net.minecraft.network.PacketBuffer
import net.minecraft.util.ResourceLocation
import net.minecraftforge.fml.LogicalSide
import net.minecraftforge.fml.network.NetworkEvent
import net.minecraftforge.fml.network.NetworkRegistry
import net.minecraftforge.fml.network.simple.SimpleChannel
import java.util.function.Function
import java.util.function.Supplier
import kotlin.reflect.KClass

object TCombatPacketHandler {
    private const val PROTOCOL_VERSION = "1"
    private const val CHANNEL_NAME = "tcombat_packet_handler"
    private var id = 0
    val INSTANCE: SimpleChannel = NetworkRegistry.newSimpleChannel(ResourceLocation(TCombatMain.MODID, CHANNEL_NAME), { PROTOCOL_VERSION }, { o: String -> PROTOCOL_VERSION == o }) { o: String ->
        PROTOCOL_VERSION == o
    }

    private fun <T: ITCombatPacket>registerPacket(packet: KClass<T>, decode: Function<PacketBuffer, T>, receiveSide: LogicalSide) {
        INSTANCE.registerMessage(id, packet.java, {p: T, buffer: PacketBuffer-> p.encode(buffer)}, decode, {p: T, supplier: Supplier<NetworkEvent.Context> ->
            if (supplier.get().direction.receptionSide != receiveSide)throw IllegalThreadStateException("${packet.simpleName} can't be received on ${supplier.get().direction.receptionSide}")
            p.handle(supplier)
            supplier.get().packetHandled = true
        })
        id++
    }

    fun register() {
        registerPacket(SkillRequestPacket::class, SkillRequestPacket::decode, LogicalSide.SERVER)
        registerPacket(TargetRequestPacket::class, TargetRequestPacket::decode, LogicalSide.SERVER)
        registerPacket(TargetModifyNotifyPacket.Mode::class, TargetModifyNotifyPacket.Mode::decode, LogicalSide.CLIENT)
        registerPacket(TargetModifyNotifyPacket.Targets::class, TargetModifyNotifyPacket.Targets::decode, LogicalSide.CLIENT)
        registerPacket(TargetModifyNotifyPacket.Focus::class, TargetModifyNotifyPacket.Focus::decode, LogicalSide.CLIENT)
        registerPacket(CBSizeRequestPacket::class, CBSizeRequestPacket::decode, LogicalSide.CLIENT)
        registerPacket(CBSizeAnswerPacket::class, CBSizeAnswerPacket::decode, LogicalSide.SERVER)
    }

}